### PR TITLE
Calculate mean average

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+category: ruby
+image:
+  base: '2020/calculate-a-mean-average-from-a-ruby-array'
+  alt: "Black and White Texas Instruments calculator"
+  credit:  Ray Reyes
+  source: "https://unsplash.com/photos/3xwrg7Vv6Ts"
+
+---
+
+Ruby doesn’t provide a native method to generate an average (mean) value from an array of integers. Its built-in [math](https://ruby-doc.org/core-2.7.0/Math.html) library deals with more complex calculations than that and there's no native `#average` or `#mean` method on `Array`.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: *Math (?)

Comment (nodunayo) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
Ruby doesn’t provide a native method to generate an average (mean) value from an array of integers. Its built-in [math](https://ruby-doc.org/core-2.7.0/Math.html) library focuses on more complex calculations and there's no built-in `#average` or `#mean` method on `Array`.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+This leaves us free to create our own implementation, with opportunities to shoot ourselves in the foot.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
This leaves us free to create our own implementation, bringing with that opportunities to shoot ourselves in the foot.
```

Not convinced on either mine or your version of this sentence...

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+a = [1, 2, 3, 4, 5, 6, 7, 8]
+
+a.sum(0.0) / a.size
+#=> 4.5
+```
+
+
+## Why?
+
+The preferred approach using [the `#sum` method from `Array`](https://ruby-doc.org/core-2.7.0/Array.html#method-i-sum) is many, many times faster than using an alternative approach using `inject`.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: *This approach of using....is many..times faster than using `inject.`

The fact that it is preferred is inferred, as is the fact that using `inject` is an alternative...

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+The `#sum` method was added to `Array` in Ruby 2.4, which is why you might see alternative implementations in other places on the Internet.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md:35]: *was _only_ added to...

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+In order to see the speed of the different implementations we can use the `benchmark-ips` gem, which compares the performance of different bits of code.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: *compare the performance of...we can use the `benchmark-ips` gem.

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    a.sum(0.0) / a.size
+  end
+  x.report('inject(0.0) / size') do
+    a.inject(0.0) { |result, i| result + i } / a.size
+  end
+  x.compare!
+end
+```
+
+The results (on my laptop) show a 50× speed improvement when using the native `#sum` in the implementation. This is due to it being written into the language to provide exactly this sort of performance improvement.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
The results (on my laptop) show a 50× speed improvement when using the native `#sum`. This method was built into the language to provide exactly this sort of performance improvement.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+Calculating -------------------------------------
+   sum(0.0) / size  680.425k (± 6.5%) i/s -  3.432M in 5.06s
+inject(0.0) / size   13.513k (± 5.0%) i/s - 67.586k in 5.01s
+
+Comparison:
+   sum(0.0) / size: 680425.2 i/s
+inject(0.0) / size:  13512.7 i/s - 50.35x  slower
+```
+
+Although there are variations using `#sum` and then `#size` that perform similarly, the most important difference is clearly the use of the native `#sum` method.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
Although there are variations of using `#sum` and then `#size` that perform similarly, the biggest performance win is choosing to use the native `#sum` method in the first place.
```

Or something like that..

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+The benchmarking code for a wider variety of implementations...
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: Colon at the end here, not ellipses.

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+ inject(0) / size.to_f:  18740.7 i/s - 35.66x slower
+ inject(0).to_f / size:  18320.1 i/s - 36.47x slower
+  inject(0).fdiv(size):  18082.6 i/s - 36.95x slower
+    inject(0.0) / size:  15264.0 i/s - 43.78x slower
+```
+
+
+## Anything else?
+
+Rails’ Active Record does include an [`#average` method](https://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-average). It’s used to perform calculations on numerical columns in the database directly in SQL.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md:128]: Technically this should be "Rails's" even though, for a long time, stylistically, I'd write the way you did: https://grammar.yourdictionary.com/punctuation/apostrophe-rules.html

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+You’re probably best to use that if that is your specific use case, creating Active Record models and then iterating over them in ruby will nearly always be slower.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: This sentence reads rather confusedly.

**Review by nodunayo — DISMISSED** [review]

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 2 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
This leaves us free to create our own implementation, bringing the opportunity to shoot ourselves in the foot.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 3 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
This approach using [the `#sum` method from `Array`](https://ruby-doc.org/core-2.7.0/Array.html#method-i-sum) is many, many times faster than using an alternative approach using `inject`.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 5 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
In order to compare the performance of the different implementations we can use the `benchmark-ips` gem.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 8 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
The benchmarking code for a wider variety of implementations:
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 10 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
If that is your specific use case, you should definitely use that method. Creating Active Record models and then iterating over them in ruby will nearly always be slower.
```

**Comment by andycroll** [comment]
Boom.

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+category: ruby
+image:
+  base: '2020/calculate-a-mean-average-from-a-ruby-array'
+  alt: "Black and White Texas Instruments calculator"
+  credit:  Ray Reyes
+  source: "https://unsplash.com/photos/3xwrg7Vv6Ts"
+
+---
+
+Ruby doesn’t provide a native method to generate an average (mean) value from an array of integers. Its built-in [math](https://ruby-doc.org/core-2.7.0/Math.html) library focuses on more complex calculations and there's no built-in `#average` or `#mean` method on `Array`.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: Well, 'maths' then, surely? As we ~Europeans~ Brits don't say 'math'? (But this is also not a big deal...)

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+This leaves us free to create our own implementation, bringing the opportunity to shoot ourselves in the foot.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: Maybe 'introducing' the opportunity...?

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+a = [1, 2, 3, 4, 5, 6, 7, 8]
+
+a.sum(0.0) / a.size
+#=> 4.5
+```
+
+
+## Why?
+
+This approach using [the `#sum` method from `Array`](https://ruby-doc.org/core-2.7.0/Array.html#method-i-sum) is many, many times faster than using an alternative approach using `inject`.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: *Using the `#sum` method is....

Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
Using [the `#sum` method from `Array`](https://ruby-doc.org/core-2.7.0/Array.html#method-i-sum) is many, many times faster than using the alternative, `inject`.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  inject(0).fdiv(size):  18082.6 i/s - 36.95x slower
+    inject(0.0) / size:  15264.0 i/s - 43.78x slower
+```
+
+
+## Anything else?
+
+Rails’ Active Record does include an [`#average` method](https://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-average). It’s used to perform calculations on numerical columns in the database directly in SQL.
+
+If that is your specific use case, you should definitely use that method. Creating Active Record models and then iterating over them in ruby will nearly always be slower.
```
Comment (nodunayo) [inline _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: *Ruby

Comment (nodunayo) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]: When you say 'creating Active Record' models, what you mean? I.e. instantiating the objects within an Array to them perform the sum and divide?

**Review by nodunayo — DISMISSED** [review]

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 11 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
Ruby doesn’t provide a native method to generate an average (mean) value from an array of integers. Its built-in [`Math`](https://ruby-doc.org/core-2.7.0/Math.html) library focuses on more complex calculations and there's no built-in `#average` or `#mean` method on `Array`.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 12 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
This leaves us free to create our own implementation, introducing the opportunity to shoot ourselves in the foot.
```

File: _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md (hunk 14 above)
Comment (andycroll) [reply _posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md]:
```suggestion
If that is your specific use case, you should definitely use that method. Instantiating Active Record models and subsequently iterating over the Array in Ruby will nearly always be slower.
```

**Pull request merged**